File: leetcode/0047_permutations_ii.cs
// https://leetcode.com/problems/permutations-ii/description/

public class Solution {
    public IList<IList<int>> PermuteUnique(int[] nums) {
        Array.Sort(nums);
        List<IList<int>> r = new List<IList<int>>{nums.ToArray()};
        
        for ( ; ; )
        {
            int i = nums.Length - 2;
            while (i >= 0 && nums[i] >= nums[i+1]) --i;
            if (i < 0) break;
            
            int j = nums.Length - 1;
            while (nums[j] <= nums[i]) --j;
            
            int t = nums[i];
            nums[i] = nums[j];
            nums[j] = t;
            
            Array.Reverse(nums, i + 1, nums.Length - i - 1);
            r.Add(nums.ToArray());
        }
        
        return r;
    }
}

File: leetcode/0047_permutations_ii.cpp
class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> r;
        sort(nums.begin(), nums.end());
        r.push_back(nums);
        while (next_perm(nums.begin(), nums.end()))
            r.push_back(nums);
        return r;
    }
    
    template <typename I> bool next_perm(I pb, I pe)
    {
        auto postpeak = is_so_u(make_reverse_iterator(pe), make_reverse_iterator(pb));
        if (postpeak.base() != pb)
            iter_swap(postpeak, upper_bound(make_reverse_iterator(pe), postpeak, *postpeak));
        reverse(make_reverse_iterator(pe), postpeak);
        return postpeak.base() != pb;
    }
    
    template <typename I> I is_so_u(I pb, I pe)
    {
        auto p = adj_f(pb, pe, greater<>());
        return p == pe ? p : next(p);
    }
    
    template <typename I, typename Predicate> I adj_f(I pb, I pe, Predicate pr)
    {
        while (pb != pe && !pr(*pb, *next(pb))) ++pb;
        return pb;
    }
};
